dask/array/core.py

```diff
@@ -1640,10 +1641,17 @@ def __float__(self):
     def __complex__(self):
         return self._scalarfunc(complex)
 
+    def __index__(self):
+        return self._scalarfunc(int)
+
     def __setitem__(self, key, value):
-        from .routines import where
+        if value is np.ma.masked:
+            value = np.ma.masked_all(())
+
+        ## Use the "where" method for cases when key is an Array
+        if isinstance(key, Array) and key.dtype == bool:
```

Comment:
Good point. I think this was a mistake, inspired by the comments on boolean indexing in that section.

I shall revert this, and the dtype change in `array/tests/test_slicing.py`

I notice that the old behaviour is inconsistent with numpy ... 
```python
>>> import dask.array as da
>>> import numpy as np
>>> x = np.arange(12,).reshape(3, 4)
>>> x
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])
>>> dx = da.from_array(x.copy())
>>> dx[dx[0]] = -99
>>> dx.compute()
array([[  0, -99, -99, -99],
       [  4, -99, -99, -99],
       [  8, -99, -99, -99]])
>>> x[x[0]] = -99
    ...
IndexError: index 3 is out of bounds for axis 0 with size 3
```
... but that's something for another day and another issue, if needs be (which it might not).